milli/src/update/index_documents/transform.rs

```diff
@@ -380,41 +390,111 @@ impl<'a, 'i> Transform<'a, 'i> {
         output_buffer: &mut Vec<u8>,
         field_buffer_cache: &mut Vec<(u16, Cow<[u8]>)>,
     ) -> Result<()> {
+        // store the keys and values of the json + the original obkv
+        let mut key_value: Vec<(FieldId, Cow<[u8]>)> = Vec::new();
+
         // if the primary_key is nested we need to flatten the document before being able to do anything
         let mut doc = serde_json::Map::new();
 
-        for (k, v) in obkv.iter() {
-            let key =
-                mapping.get(&k).ok_or(InternalError::FieldIdMappingMissingEntry { key: k })?;
-            let key = self.fields_ids_map.name(*key).ok_or(FieldIdMapMissingEntry::FieldId {
-                field_id: *key,
-                process: "Flatten from field mapping.",
-            })?;
-            let value =
-                serde_json::from_slice::<serde_json::Value>(v).map_err(InternalError::SerdeJson)?;
-            doc.insert(key.to_string(), value);
+        // we recreate a json containing only the fields that needs to be flattened.
+        // all the raw values get inserted directly in the `key_value` vec.
+        for (key, value) in obkv.iter() {
+            if json_depth_checker::should_flatten_from_unchecked_slice(value) {
+                let key =
+                    mapping.get(&key).ok_or(InternalError::FieldIdMappingMissingEntry { key })?;
+                let key =
+                    self.fields_ids_map.name(*key).ok_or(FieldIdMapMissingEntry::FieldId {
+                        field_id: *key,
+                        process: "Flatten from field mapping.",
+                    })?;
+                let value = serde_json::from_slice::<serde_json::Value>(value)
+                    .map_err(InternalError::SerdeJson)?;
+                doc.insert(key.to_string(), value);
+            } else {
+                key_value.push((key, value.into()));
+            }
         }
```

Comment:
Would it be possible to factorize this into a function and you the function instead?

Reply:
Yup, you are right, I missed that. Don't factorize those part then, keep them like that :)